pkg/controller/clustersync/clustersync_controller.go

```diff
@@ -0,0 +1,950 @@
+	}
+
+	if cd.DeletionTimestamp != nil {
+		logger.Debug("cluster is being deleted")
+		return reconcile.Result{}, nil
+	}
+
+	if unreachable, _ := remoteclient.Unreachable(cd); unreachable {
+		logger.Debug("cluster is unreachable")
+		return reconcile.Result{}, nil
+	}
+
+	restConfig, err := r.remoteClusterAPIClientBuilder(cd).RESTConfig()
+	if err != nil {
+		logger.WithError(err).Error("unable to get REST config")
+		return reconcile.Result{}, err
+	}
+	resourceHelper, err := r.resourceHelperBuilder(restConfig, logger)
+	if err != nil {
+		log.WithError(err).Error("cannot create helper")
+		return reconcile.Result{}, err
+	}
+
+	needToCreateClusterSync := false
+	clusterSync := &hiveintv1alpha1.ClusterSync{}
+	switch err := r.Get(context.Background(), request.NamespacedName, clusterSync); {
+	case apierrors.IsNotFound(err):
+		logger.Info("ClusterSync does not exist; will need to create")
+		needToCreateClusterSync = true
+	case err != nil:
+		logger.WithError(err).Log(controllerutils.LogLevel(err), "could not get ClusterSync")
+		return reconcile.Result{}, err
+	}
+
+	needToCreateLease := false
+	leaseKey := client.ObjectKey{
+		Namespace: cd.Namespace,
+		Name:      fmt.Sprintf("%s-sync", cd.Name),
+	}
+	lease := &hiveintv1alpha1.ClusterSyncLease{}
+	switch err := r.Get(context.Background(), leaseKey, lease); {
+	case apierrors.IsNotFound(err):
+		logger.Info("Lease for ClusterSync does not exist; will need to create")
+		needToCreateLease = true
+	case err != nil:
+		logger.WithError(err).Log(controllerutils.LogLevel(err), "could not create lease for ClusterSync")
+		return reconcile.Result{}, err
+	}
+
+	origStatus := clusterSync.Status.DeepCopy()
+
+	syncSets, err := r.getSyncSetsForClusterDeployment(cd, logger)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	selectorSyncSets, err := r.getSelectorSyncSetsForClusterDeployment(cd, logger)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	needToDoFullReapply := needToCreateClusterSync || r.timeUntilFullReapply(lease) <= 0
+	if needToDoFullReapply {
+		logger.Info("need to reapply all syncsets")
+	}
+
+	// Apply SyncSets
+	syncStatusesForSyncSets, syncSetsNeedRequeue := r.applySyncSets(
+		cd,
+		"SyncSet",
+		syncSets,
+		clusterSync.Status.SyncSets,
+		needToDoFullReapply,
+		resourceHelper,
+		logger,
+	)
+	clusterSync.Status.SyncSets = syncStatusesForSyncSets
+
+	// Apply SelectorSyncSets
+	syncStatusesForSelectorSyncSets, selectorSyncSetsNeedRequeue := r.applySyncSets(
+		cd,
+		"SelectorSyncSet",
+		selectorSyncSets,
+		clusterSync.Status.SelectorSyncSets,
+		needToDoFullReapply,
+		resourceHelper,
+		logger,
+	)
+	clusterSync.Status.SelectorSyncSets = syncStatusesForSelectorSyncSets
+
+	setFailedCondition(clusterSync)
+
+	// Create or Update the ClusterSync
+	if needToCreateClusterSync {
+		logger.Info("creating ClusterSync")
+		clusterSync.Namespace = cd.Namespace
+		clusterSync.Name = cd.Name
+		clusterSync.OwnerReferences = []metav1.OwnerReference{{
```

Comment:
(Here and L375) Would it be appropriate to use [`SetOwnerReference`](https://github.com/kubernetes-sigs/controller-runtime/blob/v0.6.0/pkg/controller/controllerutil/controllerutil.go#L95) or [`SetControllerReference`](https://github.com/kubernetes-sigs/controller-runtime/blob/v0.6.0/pkg/controller/controllerutil/controllerutil.go#L58) instead of inlining this? (The former doesn't `BlockOwnerDeletion`. The latter sets `Controller`, which you don't need (at least at the moment) but shouldn't hurt anything.)

Reply:
Neither of those are going to do quite what I want, as you point out. I do not want to set `Controller` to `true`: The `Controller` field is something that I am trying to get Hive away from using. So, my choices would be to either
(1) use `SetOwnerReference` and then manually set `BlockOwnerDeletion`
```
controllerutil.SetOwnerReference(clusterSync, lease, r.scheme)
lease.OwnerReferences[0].BlockOwnerDeletion = pointer.BoolPtr(true)
```
or (2) use `SetControllerReference` and then manually clear `Controller`
```
controllerutil.SetControllerReference(clusterSync, lease, r.scheme)
lease.OwnerReferences[0].Controller = nil
```
Both of those also require storing the scheme in the controller, too, which is not currently done.

Is the objection that the kind is hard-coded? If so, would it be acceptable to change it to the following?
```
lease.OwnerReferences = []metav1.OwnerReference{{
	APIVersion:         clusterSync.APIVersion,
	Kind:               clusterSync.Kind,
	Name:               clusterSync.Name,
	UID:                clusterSync.UID,
	BlockOwnerDeletion: pointer.BoolPtr(true),
}}
```

Reply:
I just thought it would be nice to reuse those functions since they're there. I didn't realize there was an objection to `Controller` -- why is that? The only thing it's used for is if you `EnqueueRequestForOwner`; otherwise it's ignored -- right?

Anyway, please feel free to disregard this comment. (Though I do prefer your first and third alternatives.)

Reply:
My objection to `Controller` is that it is meaningless and there can be only one owner reference that is the controller.